src/MOI_wrapper.jl

```diff
@@ -0,0 +1,1422 @@
+                          MOI.LessThan{Float64}, quadratic_le)
+@define_constraint_dual_start(MOI.ScalarQuadraticFunction{Float64},
+                          MOI.EqualTo{Float64}, quadratic_eq)
+
+function MOI.supports(::Optimizer, ::MOI.NLPBlockDualStart)
+    return true
+end
+function MOI.set(model::Optimizer, ::MOI.NLPBlockDualStart, values)
+    model.nlp_dual_start = -values
+    return
+end
+
+function MOI.set(model::Optimizer, ::MOI.NLPBlock, nlp_data::MOI.NLPBlockData)
+    model.nlp_data = nlp_data
+    return
+end
+
+function MOI.set(model::Optimizer, ::MOI.ObjectiveFunction,
+                 func::Union{MOI.SingleVariable, MOI.ScalarAffineFunction,
+                             MOI.ScalarQuadraticFunction})
+    check_inbounds(model, func)
+    model.objective = func
+    return
+end
+
+# In setting up the data for Ipopt, we order the constraints as follows:
+# - linear_le_constraints
+# - linear_ge_constraints
+# - linear_eq_constraints
+# - quadratic_le_constraints
+# - quadratic_ge_constraints
+# - quadratic_eq_constraints
+# - nonlinear constraints from nlp_data
+
+linear_le_offset(model::Optimizer) = 0
+linear_ge_offset(model::Optimizer) = length(model.linear_le_constraints)
+linear_eq_offset(model::Optimizer) = linear_ge_offset(model) + length(model.linear_ge_constraints)
+quadratic_le_offset(model::Optimizer) = linear_eq_offset(model) + length(model.linear_eq_constraints)
+quadratic_ge_offset(model::Optimizer) = quadratic_le_offset(model) + length(model.quadratic_le_constraints)
+quadratic_eq_offset(model::Optimizer) = quadratic_ge_offset(model) + length(model.quadratic_ge_constraints)
+nlp_constraint_offset(model::Optimizer) = quadratic_eq_offset(model) + length(model.quadratic_eq_constraints)
+
+# Convenience functions used only in optimize!
+
+function append_to_jacobian_sparsity!(jacobian_sparsity, aff::MOI.ScalarAffineFunction, row)
+    for term in aff.terms
+        push!(jacobian_sparsity, (row, term.variable_index.value))
+    end
+end
+
+function append_to_jacobian_sparsity!(jacobian_sparsity, quad::MOI.ScalarQuadraticFunction, row)
+    for term in quad.affine_terms
+        push!(jacobian_sparsity, (row, term.variable_index.value))
+    end
+    for term in quad.quadratic_terms
+        row_idx = term.variable_index_1
+        col_idx = term.variable_index_2
+        if row_idx == col_idx
+            push!(jacobian_sparsity, (row, row_idx.value))
+        else
+            push!(jacobian_sparsity, (row, row_idx.value))
+            push!(jacobian_sparsity, (row, col_idx.value))
+        end
+    end
+end
+
+# Refers to local variables in jacobian_structure() below.
+macro append_to_jacobian_sparsity(array_name)
+    escrow = esc(:row)
+    quote
+        for info in $(esc(array_name))
+            append_to_jacobian_sparsity!($(esc(:jacobian_sparsity)), info.func, $escrow)
+            $escrow += 1
+        end
+    end
+end
+
+function jacobian_structure(model::Optimizer)
```

Comment:
Is this needed? NLopt supports only dense jacobians.

Reply:
Yes, I followed the approach of the MPB wrapper: since MOI gives the jacobian as the nonzero values of the fixed sparsity structure of the jacobian, the indices of this sparsity need to be queried in order to know where these nonzero values should be put in the dense jacobian given by NLopt.